### PR TITLE
Feature/schedule management UI change

### PR DESCRIPTION
**Pull request by nishi7409**
## Features
- Daily scheduling (different times per day)
- Recurring scheduling (same time daily)
- Timezone support
- Auto Scaling Group integration
- LiteLLM automatic registration/deregistration
- CloudWatch Events monitoring

### Frontend
- `ScheduleConfig.tsx`: Scheduling UI with validation
- `AutoScalingConfig.tsx`: Toggle between 24/7 and scheduled operation
- `ModelManagementUtils.tsx`: Schedule display in model cards
- Added comprehensive test coverage

### Backend
- `schedule_management.py`: Core scheduling logic with ASG integration
- `schedule_monitoring.py`: CloudWatch Events processing and LiteLLM lifecycle
- `schedule_handlers.py`: State machine integration
- `domain_objects.py`: Enhanced data models

### LiteLLM Integration
- Models register with LiteLLM when scaling up
- Models deregister when scaling down
- Retry logic with exponential backoff

### Validation
- 24-hour time format validation
- 2-hour minimum schedule windows
- Timezone validation
- At least one day required for daily schedules

### UI Changes
- Simple Auto Scaling toggle
- Progressive disclosure of schedule options
- Clean timezone and time selection
- Improved model card display format

## Documentation
[UI Docs](https://zt9hggtte5.execute-api.us-west-2.amazonaws.com/LISA/config/model-management-ui.html#model-scheduling)
[API Docs](https://zt9hggtte5.execute-api.us-west-2.amazonaws.com/LISA/config/model-management-api.html#model-scheduling-admin-api)

## Videos
https://github.com/user-attachments/assets/234b2bc4-b874-474f-86cb-52cc3956cbc8

https://github.com/user-attachments/assets/d2b9476d-d3a9-4c0e-bf0c-3369b3dc6831

https://github.com/user-attachments/assets/31ebda8e-2a88-46e8-98fd-d0645386b906

### Auto Scaling Configuration
image
image
image

### Model Management Cards
image

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### Summary (generated)

# Summary of Changes

This comprehensive pull request implements schedule management functionality across the entire application stack, including frontend UI components, backend scheduling logic, infrastructure updates, and extensive test coverage.

### Schedule Management System
- **Daily Scheduling**: Configure different start/stop times for each day of the week
- **Recurring Scheduling**: Set consistent daily schedules across all days
- **Timezone Support**: Full IANA timezone support with validation
- **Auto Scaling Group Integration**: Direct ASG capacity management
- **LiteLLM Lifecycle Management**: Automatic model registration/deregistration during scaling events

### Frontend Enhancements
- New `ScheduleConfig.tsx` component with comprehensive scheduling UI
- Enhanced `AutoScalingConfig.tsx` with schedule integration
- `BedrockKBDiscoveryWizard.tsx` for Bedrock Knowledge Base setup
- Improved model card display with schedule information
- Responsive layout improvements (navigation width adjustment, dynamic input sizing)
- Validation for 24-hour time format, 2-hour minimum windows, and timezone requirements

### Backend Refactoring
- **New Modules**: `schedule_management.py`, `schedule_monitoring.py`, `schedule_handlers.py`
- **Data Model Updates**: 
  - Renamed enums: `EACH_DAY` → `DAILY`, `RECURRING_DAILY` → `RECURRING`
  - Introduced discriminated unions: `DailySchedulingConfig` and `RecurringSchedulingConfig`
  - Updated DynamoDB field names: `autoScalingGroup` → `auto_scaling_group`, `status` → `model_status`
  - Restructured configuration: `autoScalingConfig.scheduling` → `model_config.autoScalingConfig.scheduling`
- **CloudWatch Events Integration**: ASG instance lifecycle monitoring
- **Immediate Scaling**: Dynamic capacity adjustment based on current time and schedules

### Infrastructure Updates
- New Lambda functions: `ScheduleManagementLambda` and `ScheduleMonitoringLambda`
- CloudWatch Events rule for ASG lifecycle events
- Enhanced IAM permissions for Auto Scaling operations
- Node.js Lambda layer support with new `NodeLayer` construct
- Updated build configuration for Node.js layer compilation

### Bedrock Knowledge Base Enhancements
- Multiple data sources per Knowledge Base support
- Automatic collection creation per data source
- New API endpoints: `list_bedrock_knowledge_bases()`, `list_bedrock_data_sources()`
- S3 document discovery with pagination support
- Enhanced error handling and backward compatibility

### Build and Development Tools
- Refactored build scripts: `build_layer` → `build_python_layer`, `build_lambda` → `build_python_lambda`
- Updated esbuild configuration for external dependencies
- Python 3.11 specification and isort upgrade (5.13.2 → 7.0.0)
- Improved code maintainability with constant extraction

### Testing
- Comprehensive test coverage for new scheduling functionality
- Tests refactored to use direct function calls instead of Lambda invocations
- Edge case and error scenario coverage
- ASG integration and LiteLLM lifecycle management tests
- Bedrock Knowledge Base discovery and configuration tests

## Breaking Changes
- Lambda function name environment variable lookups now require explicit configuration (no default fallbacks)
- Schedule type enum values changed
- DynamoDB field naming conventions updated
- Configuration structure reorganization under `model_config`

## Backward Compatibility
- Optional scheduling fields maintain compatibility with existing configurations
- Legacy Bedrock KB configuration structures supported
- Flexible exception handling for JSON-wrapped or direct error messages

### CODE REVIEW

File: lambda/models/scheduling/schedule_monitoring.py [hunk 1]
```diff
@@ -432,6 +442,128 @@ def reset_retry_count(model_id: str) -> None:
...
+            UpdateExpression="SET litellm_id = :lid",
+            ExpressionAttributeValues={":lid": litellm_id},
+        )
+
+    except Exception as e:
+        logger.error(f"Failed to register {model_id} with LiteLLM: {e}", exc_info=True)
+
+
+def remove_litellm(model_id: str) -> None:
+    """Remove model from LiteLLM if registered"""
```
Comment (estohlmann) [inline lambda/models/scheduling/schedule_monitoring.py]: Same with this. No need to maintain all this logic a second time from the model management paths.

File: lambda/utilities/common_functions.py [hunk 2]
```diff
@@ -22,7 +22,7 @@
 from contextvars import ContextVar
 from decimal import Decimal
 from functools import cache
-from typing import Any, Callable, cast, Dict, Optional, TypeVar, Union
+from typing import Any, Callable, cast, Dict, List, Optional, TypeVar, Union
```
[the remark this answers is not in the document]
Comment (estohlmann) [reply lambda/utilities/common_functions.py]: Agreed, this should be removed.

File: lib/user-interface/react/src/components/model-management/create-model/AutoScalingConfig.tsx [hunk 3]
```diff
@@ -47,37 +76,66 @@ export function AutoScalingConfig (props: AutoScalingConfigProps) : ReactElement
...
                     <FormField
                         label='Desired Capacity'
                         description='Target number of instances to maintain. Must be between min and max capacity.'
-                        errorText={props.formErrors?.autoScalingConfig?.desiredCapacity}
+                        errorText={
+                            props.formErrors?.autoScalingConfig?.desiredCapacity ||
+                            (props.item.desiredCapacity && props.item.desiredCapacity < 1 ? 'Value must be greater than or equal to 1' :
+                                props.item.desiredCapacity && props.item.desiredCapacity < props.item.minCapacity ? `Value must be greater than or equal to Min Capacity (${props.item.minCapacity})` :
+                                    props.item.desiredCapacity && props.item.desiredCapacity > props.item.maxCapacity ? `Value must be less than or equal to Max Capacity (${props.item.maxCapacity})` :
+                                        undefined)
```
[the remark this answers is not in the document]
Comment (nishi7409) [reply lib/user-interface/react/src/components/model-management/create-model/AutoScalingConfig.tsx]: Broke it up into a function

File: lib/user-interface/react/src/components/model-management/ModelManagementUtils.tsx [hunk 4]
```diff
@@ -34,6 +34,80 @@ export const MODEL_STATUS_LOOKUP: EnumDictionary<ModelStatus, StatusIndicatorPro
     [ModelStatus.Failed]: 'error',
 };
 
+// Utility functions for schedule display
+const formatScheduleType = (model: IModel): string => {
+    const scheduling = model.autoScalingConfig?.scheduling;
+
+    if (!scheduling?.scheduleEnabled || !scheduling?.scheduleType || scheduling.scheduleType === ScheduleType.NONE) {
+        return '24/7';
```
[the remark this answers is not in the document]
Comment (nishi7409) [reply lib/user-interface/react/src/components/model-management/ModelManagementUtils.tsx]: Wording received from Cindy, updating to `Always on.`

File: lambda/utilities/common_functions.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (nishi7409) [reply lambda/utilities/common_functions.py]: Ohh I know what happened.  I implemented user_has_group_access and validate_model_access because get_model_and_validate_access uses it in /home/nishrs/LISA/lambda/models/handler/utils.py and that function is used here /home/nishrs/LISA/lambda/models/handler/schedule_handlers.py

Little did I know, what I wanted was already implemented here /home/nishrs/LISA/lambda/models/handler/utils.py

File: lib/user-interface/react/src/components/model-management/create-model/CreateModelModal.tsx [hunk 5]
```diff
@@ -252,21 +260,53 @@ export function CreateModelModal (props: CreateModelModalProps) : ReactElement {
...
-            if (updateFields.autoScalingConfig) {
-                updateRequest.autoScalingInstanceConfig = _.pickBy(updateFields.autoScalingConfig,
-                    (value) => value !== undefined
-                );
-            }
+            // Check if this is a scheduling-only update
+            const isSchedulingOnlyUpdate = updateFields.autoScalingConfig?.scheduling &&
+                Object.keys(updateFields).length === 2 && // modelId + autoScalingConfig
+                Object.keys(updateFields.autoScalingConfig).length === 1 && // only scheduling
+                Object.keys(updateFields.autoScalingConfig)[0] === 'scheduling';
```
[the remark this answers is not in the document]
Comment (nishi7409) [reply lib/user-interface/react/src/components/model-management/create-model/CreateModelModal.tsx]: Fixed in latest deployment

**Comment by nishi7409** [comment]
After looking through all scheduling code, I have found some dead code that doesn't serve a useful purpose-- primarily speaking of `handle_failed_scaling`. This adds complexity without solving problems, most ASG failures need human intervention. AWS ASGs already handles what can be automatically retried; other actions like EBS volume increases (or similar) must be handled by a user.

File: lambda/models/scheduling/schedule_management.py [hunk 6]
```diff
@@ -614,6 +611,121 @@ def cleanup_scheduled_actions_by_name_pattern(auto_scaling_group: str, model_id:
...
+        daily_schedule.thursday,  # 3: Thursday
+        daily_schedule.friday,  # 4: Friday
+        daily_schedule.saturday,  # 5: Saturday
+        daily_schedule.sunday,  # 6: Sunday
+    ]
+
+    # Check today first
+    today_schedule = day_schedules[current_weekday]
+    if today_schedule:
+        next_action = _get_next_action_for_day(today_schedule, now, tz, 0)
```
Comment (dustins) [inline lambda/models/scheduling/schedule_management.py]: For future design, since `today_schedule` I believe is a `DaySchedule`, it would be a little nicer to have this functionality on that object instead of a function. That concept applies to some other places too.

**Review by estohlmann — APPROVED** [review]

File: lambda/models/scheduling/schedule_management.py [hunk 7]
```diff
@@ -119,21 +123,33 @@ def update_schedule(event: Dict[str, Any]) -> Dict[str, Any]:
 
     except Exception as e:
         logger.error(f"Failed to update schedule for model {model_id}: {str(e)}")
-        raise ScheduleManagementError(f"Failed to update schedule: {str(e)}")
+        raise RuntimeError(f"Failed to update schedule: {str(e)}")
```
Comment (bedanley) [inline lambda/models/scheduling/schedule_management.py]: Why did you change from a specific to a generic error? I'd avoid RuntimeError unless it has to do with python directly.
